elastic/elastic-agent
# Build darwin/aarch64 and universal binary

## What does this PR do?

It enables the Elastic-Agent and Beats to be built for `darwin/arm64` and generates a universal binary (`darwin/universal`).

The [universal binary](https://developer.apple.com/documentation/apple-silicon/building-a-universal-macos-binary) is the merge of the `darwin/arm64` with `darwin/amd64`. Therefore it requires to first build the  `darwin/arm64` and `darwin/amd64` and them _"merge"_ them together into the `darwin/universal`. Also `darwin/universal` isn't a valid platform or OS/ARCH pair, therefore it isn't recognised by the Go compiler. Given that, the choice was to add `darwin/arm64` as a new platform for the Elastic-Agent and when ever both platforms (`darwin/arm64` and `darwin/amd64`) are built, also generate and package the universal binary.
Even though  it's ~quite~ an implicit behaviour, the alternative would also require a lot of changes and exceptions to have our build process recognising `darwin/universal`, but compiling `darwin/arm64` and `darwin/amd64` and merging them. Also the current process makes available all three of them, allowing the user to choose which fits better their needs. The drawback of the universal binary is its size, the double of the single binaries.

A new target, `BuildDarwinUniversal`  was added on the `x-pack`s `magefile`s required by the Elastic-Agent.

Also fixes some small typos, redundant information on the help command and remove the use of some deprecated functions.

## Why is it important?

We need to support Mac with M1 chips

## Checklist




## Author's Checklist


- [x] build darwin/arm64 binary
- [x] build universal 2 binary
- [x] use mage to automatically build the universal binary
- [ ] don't forget beats PR: https://github.com/elastic/beats/issues/29582

## How to test this PR locally
1. Set up the [beats repo](https://github.com/elastic/beats/):
   - have the beats repo on the same level as the `elastic-agent` repo
   - set the beats repo to the branch `26683-agent-arm64` while https://github.com/elastic/beats/pull/29585 isn't merged
1. Build the elastic-agent for the new platforms:
   - **darwin/arm64**: `PLATFORMS="darwin/arm64" PACKAGES="tar.gz" mage package`
   - **darwin/universal**: `PLATFORMS="darwin/arm64,darwin/amd64" PACKAGES="tar.gz" mage package`

3. Run the elastic-agent

####**Watch out:**
The elastic-agent tries to download any binary it might need to comply with the assigned policy. However none of the new binaries from this PR are present on the artifact API and other applications (e.g. fleet-server and elastic-endpoint) also aren't publishing `darwin/arm64` versions. Therefore if the elastic-agent tries to download them, it'll fail. This is the expected behavior and this PR does not address it, see *Out of Scope*.

## Out of scope:
 - Handle downloading binaries that does not publish `darwin-universal` artifacts on the artifact API
 - Handle upgrading a `darwin/amd64` agent running under Rosetta2 on a Mac with M1 chip. This will be addressed on another PR.

## Related issues

- Relates elastic/elastic-agent#151
- Closes elastic/elastic-agent#204
